File: shader.py
# # Requirements # #
# pip install glfw pyopengl pyrr pillow

import sys

import glfw
from OpenGL.GL import *
import OpenGL.GL.shaders
import numpy
from PIL import Image


# Initialize glfw
if not glfw.init():
    print("Couldn't initialize GLFW")
    exit(1)

# Create window
window = glfw.create_window(1, 1, "My OpenGL window", None, None)  # Size (1, 1) for show nothing in window
# window = glfw.create_window(800, 600, "My OpenGL window", None, None)

# Terminate if any issue
if not window:
    glfw.terminate()
    print("Couldn't create window")
    exit(1)

# Set context to window
glfw.make_context_current(window)

# Initial data
quad_data = [
#   positions     texture coordinates
    -1., -1., 0., 0., 0.,
     1., -1., 0., 1., 0.,
     1.,  1., 0., 1., 1.,
    -1.,  1., 0., 0., 1.
]
quad = numpy.array(quad_data, dtype=numpy.float32)
# Vertices indices order
indices_data = [
    0, 1, 2,
    2, 3, 0
]
indices = numpy.array(indices_data, dtype=numpy.uint32)

# Vertex shader
vertex_shader = """
#version 330
layout(location = 0) in vec3 position;
layout(location = 1) in vec2 inTexCoords;
out vec2 outTexCoords;

void main() {
    gl_Position = vec4(position, 1.0);
    outTexCoords = inTexCoords;
}
"""

# Fragment shader
fragment_shader = """
#version 330

uniform sampler2D source;
in vec2 outTexCoords;

float intensityFactor = 2.;

void main() {
    ivec2 textureSize2d = textureSize(source, 0); // Width and height of texture image

    // coloring
    //vec3 outColor = vec3(0.3, 0.1, 0.2);
    //gl_FragColor = texture(source, outTexCoords) * vec4(outColor, 1.0f);

    // inversion
    //gl_FragColor = vec4(1., 1., 1., 2.) - texture(source, outTexCoords);

    // RGB
    //vec2 uv = gl_FragCoord.xy / textureSize2d;
    //vec3 color;
    //if (uv.x < 1./3.) {
    //    color = vec3(1., 0., 0.);
    //} else if (uv.x < 2./3.) {
    //    color = vec3(0., 1., 0.);
    //} else {
    //    color = vec3(0., 0., 1.);
    //}
    //gl_FragColor = vec4(texture(source, outTexCoords).rgb * color, 1.);

    // remove body color
    // TODO: smooth
    //vec4 c = texture(source, outTexCoords);
    //float threshold = 60. / 255.;
    //if (abs(c.r - 74./255.) < threshold && abs(c.g - 38./255.) < threshold && abs(c.b - 26./255.) < threshold) {
    //    gl_FragColor = vec4(vec3((c.r + c.g + c.b) / 3.), 1.);
    //} else {
    //    gl_FragColor = c;
    //}

    // mirror horizontally
    if (gl_FragCoord.x / textureSize2d.x < 0.5) {
        gl_FragColor = texture(source, vec2(textureSize2d.x - outTexCoords.x, outTexCoords.y));
    } else {
        gl_FragColor = texture(source, outTexCoords);
    }

    //gl_FragColor = vec4((k1*t1 + k2*t2 + k3*t3)/(k1 + k2 + k3), 1.0f);
    //gl_FragColor = vec4(s00, 1.0f);
}
"""

# Compile shaders
try:
    shader = OpenGL.GL.shaders.compileProgram(
        OpenGL.GL.shaders.compileShader(vertex_shader, GL_VERTEX_SHADER),
        OpenGL.GL.shaders.compileShader(fragment_shader, GL_FRAGMENT_SHADER)
    )
except OpenGL.GL.shaders.ShaderCompilationError as e:
    print("Error compiling shader:")
    for x in e.args:
        print(x)
    exit(1)


# VBO
vertex_buffer_object = glGenBuffers(1)
glBindBuffer(GL_ARRAY_BUFFER, vertex_buffer_object)
glBufferData(GL_ARRAY_BUFFER, quad.itemsize * len(quad), quad, GL_STATIC_DRAW)


# EBO
entity_buffer_object = glGenBuffers(1)
glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, entity_buffer_object)
glBufferData(GL_ELEMENT_ARRAY_BUFFER, indices.itemsize * len(indices), indices, GL_STATIC_DRAW)

# Configure positions of initial data
glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, quad.itemsize * 5, ctypes.c_void_p(0))
glEnableVertexAttribArray(0)

# Configure texture coordinates of initial data
glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, quad.itemsize * 5, ctypes.c_void_p(12))
glEnableVertexAttribArray(1)


# Texture
texture = glGenTextures(1)
# Bind texture
glBindTexture(GL_TEXTURE_2D, texture)
# Texture wrapping params
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
# Texture filtering params
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)

# Open image
image = Image.open(sys.argv[1])
img_data = image.convert("RGBA").tobytes()
glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, image.width, image.height, 0, GL_RGBA, GL_UNSIGNED_BYTE, img_data)

# Create render buffer with size (image.width x image.height)
rb_obj = glGenRenderbuffers(1)
glBindRenderbuffer(GL_RENDERBUFFER, rb_obj)
glRenderbufferStorage(GL_RENDERBUFFER, GL_RGBA, image.width, image.height)

# Create frame buffer
fb_obj = glGenFramebuffers(1)
glBindFramebuffer(GL_FRAMEBUFFER, fb_obj)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, rb_obj)

# Check frame buffer (that simple buffer should not be an issue)
status = glCheckFramebufferStatus(GL_FRAMEBUFFER)
if status != GL_FRAMEBUFFER_COMPLETE:
    print("incomplete framebuffer object")
    exit(1)

# Install program
glUseProgram(shader)

# Bind framebuffer and set viewport size
glBindFramebuffer(GL_FRAMEBUFFER, fb_obj)
glViewport(0, 0, image.width, image.height)

# Draw the quad which covers the entire viewport
glDrawElements(GL_TRIANGLES, 6, GL_UNSIGNED_INT, None)

# Read the data and create the image
image_buffer = glReadPixels(0, 0, image.width, image.height, GL_RGBA, GL_UNSIGNED_BYTE)
image_out = numpy.frombuffer(image_buffer, dtype=numpy.uint8).reshape(image.height, image.width, 4)
img = Image.fromarray(image_out, 'RGBA')
img.save(r"image_out.png")

File: shader.go
package fimgs

import (
	"fmt"
	"image"
	"image/draw"
	"os"
	"reflect"
	"strings"
	"unsafe"

	"github.com/go-gl/gl/v4.1-core/gl"
	"github.com/go-gl/glfw/v3.3/glfw"
)

func compileShader(source string, shaderType uint32) (uint32, error) {
	shader := gl.CreateShader(shaderType)

	csources, free := gl.Strs(source)
	gl.ShaderSource(shader, 1, csources, nil)
	free()
	gl.CompileShader(shader)

	var status int32
	gl.GetShaderiv(shader, gl.COMPILE_STATUS, &status)
	if status == gl.FALSE {
		var logLength int32
		gl.GetShaderiv(shader, gl.INFO_LOG_LENGTH, &logLength)

		log := strings.Repeat("\x00", int(logLength+1))
		gl.GetShaderInfoLog(shader, logLength, nil, gl.Str(log))

		return 0, fmt.Errorf("failed to compile shader source:\n%v\nerror:\n%v", source, log)
	}
	return shader, nil
}

func newTexture(file string) (int, int, error) {
	imgFile, err := os.Open(file)
	if err != nil {
		return 0, 0, fmt.Errorf("texture %q not found on disk: %v", file, err)
	}
	img, _, err := image.Decode(imgFile)
	if err != nil {
		return 0, 0, err
	}

	rgba := image.NewRGBA(img.Bounds())
	if rgba.Stride != rgba.Rect.Size().X*4 {
		return 0, 0, fmt.Errorf("unsupported stride")
	}
	draw.Draw(rgba, rgba.Bounds(), img, image.Point{0, 0}, draw.Src)

	var texture uint32
	gl.GenTextures(1, &texture)
	gl.ActiveTexture(gl.TEXTURE0)
	gl.BindTexture(gl.TEXTURE_2D, texture)
	gl.TexParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR)
	gl.TexParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR)
	gl.TexParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE)
	gl.TexParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE)
	gl.TexImage2D(
		gl.TEXTURE_2D,
		0,
		gl.RGBA,
		int32(rgba.Rect.Size().X),
		int32(rgba.Rect.Size().Y),
		0,
		gl.RGBA,
		gl.UNSIGNED_BYTE,
		gl.Ptr(rgba.Pix),
	)
	return rgba.Rect.Size().X, rgba.Rect.Size().Y, nil
}

// requires libgl1-mesa-dev, xorg-dev packages
func ShaderFilter(sourceImageFilename, resultImageFilename, fragmentShaderSource string) error {
	if err := glfw.Init(); err != nil {
		return fmt.Errorf("couldn't initialize glfw: %q", err)
	}
	defer glfw.Terminate()

	// FIX: panics sometimes in webserver mode
	window, err := glfw.CreateWindow(1, 1, "Thou shalt not exist", nil, nil) // Size (1, 1) for show nothing in window
	if err != nil {
		return fmt.Errorf("couldn't create window: %q", err)
	}

	// Set context to window
	window.MakeContextCurrent()

	if err := gl.Init(); err != nil {
		return fmt.Errorf("couldn't initialize glow: %q", err)
	}

	// Initial data
	quad := []float32{
		// [x, y, z=0] positions [u=(x+1)/2, v=(y+1)/2] texture coordinates
		-1., -1., 0., 0., 0.,
		1., -1., 0., 1., 0.,
		1., 1., 0., 1., 1.,
		-1., 1., 0., 0., 1.,
	}
	// Vertices indices order
	indices := []uint32{
		0, 1, 2,
		2, 3, 0,
	}

	// Vertex shader
	vertex_shader := `#version 330
layout(location = 0) in vec3 position;
layout(location = 1) in vec2 inTexCoords;
out vec2 outTexCoords;

void main() {
    gl_Position = vec4(position, 1.0);
    outTexCoords = inTexCoords;
}`

	// Compile shaders
	vertexShader, err := compileShader(vertex_shader, gl.VERTEX_SHADER)
	if err != nil {
		return fmt.Errorf("error compiling vertex shader:\n%q", err)
	}
	fragmentShader, err := compileShader(fragmentShaderSource, gl.FRAGMENT_SHADER)
	if err != nil {
		return fmt.Errorf("error compiling fragment shader:\n%s", err)
	}
	program := gl.CreateProgram()
	gl.AttachShader(program, vertexShader)
	gl.AttachShader(program, fragmentShader)
	gl.LinkProgram(program)

	var vertex_buffer_object uint32
	gl.GenBuffers(1, &vertex_buffer_object)
	gl.BindBuffer(gl.ARRAY_BUFFER, vertex_buffer_object)
	gl.BufferData(gl.ARRAY_BUFFER, len(quad)*4, gl.Ptr(quad), gl.STATIC_DRAW)

	// EBO
	var entity_buffer_object uint32
	gl.GenBuffers(1, &entity_buffer_object)
	gl.BindBuffer(gl.ELEMENT_ARRAY_BUFFER, entity_buffer_object)
	gl.BufferData(gl.ELEMENT_ARRAY_BUFFER, len(indices)*4, gl.Ptr(indices), gl.STATIC_DRAW)

	// Configure positions of initial data
	gl.VertexAttribPointerWithOffset(0, 3, gl.FLOAT, false, 4*5, 0)
	gl.EnableVertexAttribArray(0)

	// Configure texture coordinates of initial data
	gl.VertexAttribPointerWithOffset(1, 2, gl.FLOAT, false, 4*5, 12)
	gl.EnableVertexAttribArray(1)

	imageWidth, imageHeight, err := newTexture(sourceImageFilename)
	if err != nil {
		return fmt.Errorf("error loading texture: %q", err)
	}

	// Create render buffer with size (image.width x image.height)
	var rb_obj uint32
	gl.GenRenderbuffers(1, &rb_obj)
	gl.BindRenderbuffer(gl.RENDERBUFFER, rb_obj)
	gl.RenderbufferStorage(gl.RENDERBUFFER, gl.RGBA, int32(imageWidth), int32(imageHeight))

	// Create frame buffer
	var fb_obj uint32
	gl.GenFramebuffers(1, &fb_obj)
	gl.BindFramebuffer(gl.FRAMEBUFFER, fb_obj)
	gl.FramebufferRenderbuffer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.RENDERBUFFER, rb_obj)

	// FIX: sometimes fail with 0
	if status := gl.CheckFramebufferStatus(gl.FRAMEBUFFER); status != gl.FRAMEBUFFER_COMPLETE {
		return fmt.Errorf("incomplete framebuffer object, status is %d, gl error is %q", status, gl.GetError())
	}

	// Install program
	gl.UseProgram(program)

	// Bind framebuffer and set viewport size
	gl.BindFramebuffer(gl.FRAMEBUFFER, fb_obj)
	gl.Viewport(0, 0, int32(imageWidth), int32(imageHeight))

	gl.DrawElements(gl.TRIANGLES, 6, gl.UNSIGNED_INT, nil)

	var data []uint8 = make([]uint8, 4*imageHeight*imageWidth)
	gl.ReadPixels(0, 0, int32(imageWidth), int32(imageHeight), gl.RGBA, gl.UNSIGNED_BYTE, unsafe.Pointer((*reflect.SliceHeader)(unsafe.Pointer(&data)).Data))

	image_out := image.RGBA{
		Pix:    data,
		Stride: imageWidth * 4,
		Rect:   image.Rect(0, 0, imageWidth, imageHeight),
	}
	if err = saveImage(&image_out, resultImageFilename); err != nil {
		return fmt.Errorf("error saving file: %q", err)
	}
	return nil
}
